### PR TITLE
fix: multitenancy changes

### PR DESCRIPTION
**Pull request by sattvikc**
## Summary of change

Multitenancy recipe APIs and impl

## Related issues
- https://github.com/supertokens/supertokens-core/pull/555
- https://github.com/supertokens/supertokens-core/issues/546

## Documentation changes

(If relevant, please create a PR in our [docs repo](https://github.com/supertokens/docs), or create a checklist here
highlighting the necessary changes)

## Checklist for important updates

- [ ] Changelog has been updated
    - [ ] If there are any db schema changes, mention those changes clearly
- [ ] `coreDriverInterfaceSupported.json` file has been updated (if needed)
- [ ] `pluginInterfaceSupported.json` file has been updated (if needed)
- [ ] Changes to the version if needed
    - In `build.gradle`
- [ ] If added a new paid feature, edit the `getPaidFeatureStats` function in FeatureFlag.java file
- [ ] Had installed and ran the pre-commit hook
- [ ] If there are new dependencies that have been added in `build.gradle`, please make sure to add them
  in `implementationDependencies.json`.
- [ ] Issue this PR against the latest non released version branch.
    - To know which one it is, run find the latest released tag (`git tag`) in the format `vX.Y.Z`, and then find the
      latest branch (`git branch --all`) whose `X.Y` is greater than the latest released tag.
    - If no such branch exists, then create one from the latest released branch.

## Remaining TODOs for this PR

- [ ] Disassociate user from tenant (end to end impl)
- [ ] Tests for all the newly added APIs

### CODE REVIEW

File: src/main/java/io/supertokens/webserver/api/multitenancy/BaseCreateOrUpdate.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+
+        } catch (DeletionInProgressException | CannotModifyBaseConfigException | BadPermissionException |
+                 StorageQueryException | FeatureNotEnabledException |
+                 InvalidProviderConfigException | TenantOrAppNotFoundException e) {
+            throw new ServletException(e);
+        } catch (InvalidConfigException e) {
+            JsonObject result = new JsonObject();
+            result.addProperty("status", "INVALID_CORE_CONFIG_ERROR");
+            result.addProperty("message", e.getMessage());
+            super.sendJsonResponse(200, result, resp);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/multitenancy/BaseCreateOrUpdate.java:123]: bac request response with an appropriate message

File: src/main/java/io/supertokens/webserver/api/multitenancy/BaseRemove.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    protected void handle(TenantIdentifier sourceTenantIdentifier, TenantIdentifier targetTenantIdentifier, HttpServletResponse resp)
+            throws IOException, ServletException {
+        TenantConfig tenantConfig = Multitenancy.getTenantInfo(main, targetTenantIdentifier);
+
+        boolean didExist = false;
+
+        try {
+            Multitenancy.checkSourceAndTargetTenantForCreateOrUpdate(sourceTenantIdentifier, targetTenantIdentifier);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/multitenancy/BaseRemove.java:49]: name is bad

File: src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateAppAPI.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
+
+        String appId = null;
+        Boolean emailPasswordEnabled = null;
+        Boolean thirdPartyEnabled = null;
+        Boolean passwordlessEnabled = null;
+        JsonObject coreConfig = null;
+
+        if (input.has("appId")) {
+            appId = InputParser.parseStringOrThrowError(input, "appId", true);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateAppAPI.java:56]: trim

File: src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateAppAPI.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    @Override
+    protected void doPut(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
+        JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
+
+        String appId = null;
+        Boolean emailPasswordEnabled = null;
+        Boolean thirdPartyEnabled = null;
+        Boolean passwordlessEnabled = null;
+        JsonObject coreConfig = null;
+
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateAppAPI.java:54]: redo all input parsing in all the APIs:
- do not check for if it exists, just pass in `true` as the last arg
- input normalisation (trimming)

File: src/main/java/io/supertokens/webserver/api/multitenancy/RemoveAppAPI.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    @Override
+    public String getPath() {
+        return "/recipe/multitenancy/app/remove";
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
+        JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
+
+        String appId = InputParser.parseStringOrThrowError(input, "appId", false);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/multitenancy/RemoveAppAPI.java:48]: fix all the parseStringOrThrowError in all the new APIs

File: src/main/java/io/supertokens/webserver/api/multitenancy/thirdparty/CreateOrUpdateThirdPartyConfigAPI.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+            if (!tenantIdentifier.getTenantId().equals(TenantIdentifier.DEFAULT_TENANT_ID)) {
+                sourceTenant = new TenantIdentifier(
+                        tenantIdentifier.getConnectionUriDomain(), tenantIdentifier.getAppId(), null);
+            } else if (!tenantIdentifier.getAppId().equals(TenantIdentifier.DEFAULT_APP_ID)) {
+                sourceTenant = new TenantIdentifier(
+                        tenantIdentifier.getConnectionUriDomain(), null, null);
+            } else {
+                sourceTenant = new TenantIdentifier(null, null, null);
+            }
+            Multitenancy.addNewOrUpdateAppOrTenant(main, sourceTenant, updatedConfig);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/multitenancy/thirdparty/CreateOrUpdateThirdPartyConfigAPI.java:105]: remove sourceTenant from addNewOrUpdateAppOrTenant and do the permission check elsewhere (in the API layer)

File: src/main/java/io/supertokens/webserver/api/multitenancy/thirdparty/CreateOrUpdateThirdPartyConfigAPI.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+        } catch (TenantOrAppNotFoundException | StorageQueryException | FeatureNotEnabledException |
+                 InvalidConfigException | CannotModifyBaseConfigException | BadPermissionException e) {
+            throw new ServletException(e);
+        } catch (InvalidProviderConfigException | DeletionInProgressException e) {
+            throw new RuntimeException(e);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/multitenancy/thirdparty/CreateOrUpdateThirdPartyConfigAPI.java:111]: InvalidProviderConfigException -> bad input with clear message

File: src/main/java/io/supertokens/webserver/api/multitenancy/thirdparty/RemoveThirdPartyConfigAPI.java [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+
+            JsonObject result = new JsonObject();
+            result.addProperty("status", "OK");
+            result.addProperty("didConfigExist", found);
+
+        } catch (TenantOrAppNotFoundException | StorageQueryException | FeatureNotEnabledException |
+                 InvalidConfigException | CannotModifyBaseConfigException | BadPermissionException e) {
+            throw new ServletException(e);
+        } catch (InvalidProviderConfigException | DeletionInProgressException e) {
+            throw new RuntimeException(e);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/multitenancy/thirdparty/RemoveThirdPartyConfigAPI.java:112]: InvalidProviderConfigException -> bad input

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/multitenancy/Multitenancy.java [hunk 9]
```diff
@@ -263,25 +268,36 @@ public static void deleteConnectionUriDomain(TenantIdentifier tenantIdentifier,
...
+        } catch (TenantOrAppNotFoundException e) {
+            // we ignore this since it may have been that past deletion attempt deleted this successfully,
+            // but not from the main table.
+        }
+        StorageLayer.getMultitenancyStorage(main).deleteConnectionUriDomainInfoInBaseStorage(connectionUriDomain);
         MultitenancyHelper.getInstance(main).refreshTenantsInCoreIfRequired(true);
     }
 
     // TODO: removeUserIdFromTenant
     // TODO: removeRoleFromTenant
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/multitenancy/Multitenancy.java:287]: Removed

File: src/main/java/io/supertokens/webserver/api/multitenancy/AssociateUserToTenantAPI.java [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+
+    @Override
+    public String getPath() {
+        return "/recipe/multitenancy/tenant/user";
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
+        JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
+        String userId = InputParser.parseStringOrThrowError(input, "userId", false);
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/multitenancy/AssociateUserToTenantAPI.java:56]: added trim in all the APIs

File: src/main/java/io/supertokens/webserver/api/multitenancy/DisassociateUserFromTenant.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+    @Override
+    public String getPath() {
+        return "/recipe/multitenancy/tenant/user/remove";
+    }
+
+    @Override
+    protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
+        JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
+        String userId = InputParser.parseStringOrThrowError(input, "userId", false);
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/multitenancy/DisassociateUserFromTenant.java:52]: Done

File: src/main/java/io/supertokens/webserver/api/multitenancy/RemoveAppAPI.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/multitenancy/RemoveAppAPI.java:48]: Done

File: src/main/java/io/supertokens/webserver/api/multitenancy/ListAppsAPI.java [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+
+            Map<String, List<TenantConfig>> appsToTenants = new HashMap<>();
+            for (TenantConfig tenantConfig : tenantConfigs) {
+                if (!appsToTenants.containsKey(tenantConfig.tenantIdentifier.getAppId())) {
+                    appsToTenants.put(tenantConfig.tenantIdentifier.getAppId(), new ArrayList<>());
+                }
+                appsToTenants.get(tenantConfig.tenantIdentifier.getAppId()).add(tenantConfig);
+            }
+
+            Multitenancy.getAllTenantsForApp(tenantIdentifier.toAppIdentifier(), main);
```
[the remark this answers is not in the document]
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/multitenancy/ListAppsAPI.java:74]: copy mistake. removed it.

File: src/main/java/io/supertokens/webserver/api/multitenancy/BaseRemove.java (hunk 2 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/multitenancy/BaseRemove.java:49]: Done

File: src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateAppAPI.java (hunk 3 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateAppAPI.java:56]: Done

File: src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateAppAPI.java (hunk 4 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateAppAPI.java:54]: Done

File: src/main/java/io/supertokens/webserver/api/multitenancy/RemoveAppAPI.java (hunk 5 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/multitenancy/RemoveAppAPI.java:48]: Done

File: src/main/java/io/supertokens/webserver/api/multitenancy/thirdparty/RemoveThirdPartyConfigAPI.java (hunk 8 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/multitenancy/thirdparty/RemoveThirdPartyConfigAPI.java:112]: Done

File: src/main/java/io/supertokens/webserver/InputParser.java [hunk 13]
```diff
@@ -104,7 +104,7 @@ public static Long getLongQueryParamOrThrowError(HttpServletRequest request, Str
     public static JsonObject parseJsonObjectOrThrowError(JsonObject element, String fieldName, boolean nullable)
             throws ServletException {
         try {
-            if (nullable && element.get(fieldName) == null) {
+            if (nullable && (element.get(fieldName) == null || element.get(fieldName).isJsonNull())) {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/InputParser.java:107]: revert changes

File: src/main/java/io/supertokens/webserver/api/multitenancy/BaseRemove.java [hunk 14]
```diff
@@ -46,7 +46,7 @@ protected void handle(TenantIdentifier sourceTenantIdentifier, TenantIdentifier
         boolean didExist = false;
 
         try {
-            Multitenancy.checkSourceAndTargetTenantForCreateOrUpdate(sourceTenantIdentifier, targetTenantIdentifier);
+            Multitenancy.checkPermissionsForCreateUpdateOrDelete(main, sourceTenantIdentifier, targetTenantIdentifier);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/multitenancy/BaseRemove.java:49]: recheck deletion permissions.. what happens if i try to delete the tenant that i am querying from? Is that allowed or not?

Comment (rishabhpoddar) [reply src/main/java/io/supertokens/webserver/api/multitenancy/BaseRemove.java:49]: The permission checks should be the same as the list APIs

Comment (rishabhpoddar) [reply src/main/java/io/supertokens/webserver/api/multitenancy/BaseRemove.java:49]: You can't remove the public tenantId from the tenant remove API.

File: src/main/java/io/supertokens/webserver/api/multitenancy/RemoveTenantAPI.java [hunk 15]
```diff
@@ -45,7 +45,10 @@ public String getPath() {
     protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
         JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
 
-        String tenantId = InputParser.parseStringOrThrowError(input, "tenantId", false);
+        String tenantId = InputParser.parseStringOrThrowError(input, "tenantId", true);
```
[the remark this answers is not in the document]
Comment (rishabhpoddar) [reply src/main/java/io/supertokens/webserver/api/multitenancy/RemoveTenantAPI.java:48]: check public is not allowed

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/webserver/api/multitenancy/RemoveAppAPI.java [hunk 16]
```diff
@@ -45,18 +50,25 @@ public String getPath() {
     protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
         JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
 
-        String appId = InputParser.parseStringOrThrowError(input, "appId", true);
-        if (appId != null) {
-            appId = appId.trim();
+        String appId = InputParser.parseStringOrThrowError(input, "appId", false);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/multitenancy/RemoveAppAPI.java:53]: need to trim appId? Please make sure all APIs have this.

File: src/main/java/io/supertokens/webserver/api/multitenancy/RemoveTenantAPI.java [hunk 17]
```diff
@@ -45,20 +49,27 @@ public String getPath() {
...
+            throw new ServletException(new BadPermissionException("Cannot delete public tenant, use remove app API instead"));
         }
 
         try {
             TenantIdentifier sourceTenantIdentifier = this.getTenantIdentifierWithStorageFromRequest(req);
 
-            super.handle(
-                    sourceTenantIdentifier,
-                    new TenantIdentifier(sourceTenantIdentifier.getConnectionUriDomain(), sourceTenantIdentifier.getAppId(), tenantId),
-                    resp);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/multitenancy/RemoveTenantAPI.java]: you need to make sure that the tenant is not deleting itself?

File: src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateAppAPI.java [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+    @Override
+    public String getPath() {
+        return "/recipe/multitenancy/app";
+    }
+
+    @Override
+    protected void doPut(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
+        JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
+
+        String appId = InputParser.parseStringOrThrowError(input, "appId", true);
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateAppAPI.java:49]: In all APIs that take tenantId or appId as an input, you need to trim and lowercase it.

File: src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateAppAPI.java [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+        if (appId != null) {
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateAppAPI.java:50]: In APIs that create appId and tenantId, you need to make sure that they follow the right format (See the master checklist)

File: src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateConnectionUriDomainAPI.java [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+        return "/recipe/multitenancy/connectionuridomain";
+    }
+
+    @Override
+    protected void doPut(HttpServletRequest req, HttpServletResponse resp) throws IOException, ServletException {
+        JsonObject input = InputParser.parseJsonObjectOrThrowError(req);
+
+        String connectionUriDomain =  InputParser.parseStringOrThrowError(input, "connectionUriDomain", true);
+        if (connectionUriDomain != null) {
+            connectionUriDomain = connectionUriDomain.trim();
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateConnectionUriDomainAPI.java:50]: You need to make sure that the `connectionUriDomain` is a valid URL (it should not contain http / https)

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/multitenancy/Multitenancy.java [hunk 21]
```diff
@@ -44,32 +46,14 @@
...
-            InvalidProviderConfigException, TenantOrAppNotFoundException {
-        return addNewOrUpdateAppOrTenant(main, sourceTenant, newTenant, false);
-    }
-
-    public static boolean addNewOrUpdateAppOrTenant(Main main, TenantIdentifier sourceTenant, TenantConfig newTenant,
-                                                    boolean shouldPreventDbConfigUpdate)
-            throws DeletionInProgressException, CannotModifyBaseConfigException, BadPermissionException,
-            StorageQueryException, FeatureNotEnabledException, IOException, InvalidConfigException,
-            InvalidProviderConfigException, TenantOrAppNotFoundException {
+    public static void checkPermissionsForCreateUpdateOrDelete(Main main, TenantIdentifier sourceTenant,
```
Comment (rishabhpoddar) [inline src/main/java/io/supertokens/multitenancy/Multitenancy.java:49]: function needs to be renamed to remove the "Delete" postfix

**Review by rishabhpoddar — CHANGES_REQUESTED** [review]

File: src/main/java/io/supertokens/webserver/api/multitenancy/RemoveTenantAPI.java (hunk 17 above)
Comment (rishabhpoddar) [reply src/main/java/io/supertokens/webserver/api/multitenancy/RemoveTenantAPI.java]: Please clearly write down somewhere the exact permissions for each of the CRUD APIs for muti tenancy.. there still seems to be some misundersatnding

File: src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateAppAPI.java (hunk 18 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateAppAPI.java:49]: Done

File: src/main/java/io/supertokens/multitenancy/Multitenancy.java (hunk 21 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/multitenancy/Multitenancy.java:49]: Done

File: src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateConnectionUriDomainAPI.java (hunk 20 above)
Comment (sattvikc) [reply src/main/java/io/supertokens/webserver/api/multitenancy/CreateOrUpdateConnectionUriDomainAPI.java:50]: Done

**Review by rishabhpoddar — APPROVED** [review]

**Pull request merged**